### PR TITLE
Add KernelSpecManager.get_all_specs

### LINKED ISSUE #66: simplify get/find_kernel_specs.
listing all kernelspec (`$ jupyter kernelspec list`) make N+1 call to find_kernel_specs (one for each found kernel) 

This could be highly simplified with a "find_all_specs".

Not important, but simple for beginner, so marking as sprint friendly.

### PR DESCRIPTION
**Pull request by captainsafia**
Closes #66.

cc/ @minrk @Carreau

### CODE REVIEW

File: jupyter_client/kernelspec.py [hunk 1]
```diff
@@ -138,17 +138,10 @@ def find_kernel_specs(self):
         return d
         # TODO: Caching?
 
-    def get_kernel_spec(self, kernel_name):
-        """Returns a :class:`KernelSpec` instance for the given kernel_name.
-
-        Raises :exc:`NoSuchKernel` if the given kernel name is not found.
+    def get_kernel_spec_helper(self, kernel_name, resource_dir):
```
Comment (JanSchulz) [inline jupyter_client/kernelspec.py]: This is IMO misnamed, at least I have trouble seeing this NOT as a function, which gets a helper object, but it returns a kernel spec as far as I see it. So `_get_kernel_spec_for_name(name, dir)`?

File: jupyter_client/kernelspec.py [hunk 2]
```diff
@@ -138,17 +138,10 @@ def find_kernel_specs(self):
...
+        """ Returns a :class:`KernelSpec` instance for a given kernel_name
+        and resource_dir.
         """
-        d = self.find_kernel_specs()
-        try:
-            resource_dir = d[kernel_name.lower()]
```
Comment (JanSchulz) [inline jupyter_client/kernelspec.py]: Doesn't this make it `get_kernel_spec_helper(self, kernel_name)` as it makes the passed in `ressource_dir` argument obsolete?

**Comment by JanSchulz** [comment]
Does this also need a change in the notebook to take advantage of this new thing?

File: jupyter_client/kernelspec.py (hunk 1 above)
Comment (captainsafia) [inline jupyter_client/kernelspec.py]: I concur with what you say, @JanSchulz. Let me update the code.

**Comment by captainsafia** [comment]
Noted, not sure if it'll be include in the 4.1 release but I don't see any reason as to why it shouldn't.

File: jupyter_client/kernelspec.py [hunk 3]
```diff
@@ -161,6 +154,26 @@ def get_kernel_spec(self, kernel_name):
...
+        """
+        d = self.find_kernel_specs()
+        try:
+            resource_dir = d[kernel_name.lower()]
+        except KeyError:
+            raise NoSuchKernel(kernel_name)
+
+        return self._get_kernel_spec_by_name(kernel_name, resource_dir)
+
+    def find_all_specs(self):
```
Comment (minrk) [inline jupyter_client/kernelspec.py]: let's call this `get_all_specs`. That way `get_` returns kernelspec objects, while `find_` returns locations on disk.

And don't forget to add a docstring.

**Comment by minrk** [comment]
Thanks!

**Comment by minrk** [comment]
@captainsafia feel free to ping with a comment when you've addressed notes, so we can review/merge; we don't get notified of new commits, only comments.

Thanks again, this is great!

**Comment by captainsafia** [comment]
Yep! For some strange reason, I thought i had more work to do on this before it was merge ready but if it looks good, sweet!

**Comment by captainsafia** [comment]
I remember it now. Had to rename the `test_find_all_specs` to `test_get_all_specs` but I don't think that's too big a deal.

**Pull request merged**